File: array/15_3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // sort may infect the order of vector
		vector<vector<int>> result;
		sort(nums.begin(), nums.end());
		int len = nums.size();
		for (int i=0; i<len; )
		{
			int num = nums[i], st = i+1, end = len-1, anum = 0 - num;
			while (st < end)
			{
				if (nums[st] + nums[end] == anum)
				{
					vector<int> tpv;
					int v1 = nums[st], v2 = nums[end];
					tpv.push_back(num);
					tpv.push_back(v1);
					tpv.push_back(v2);
					result.push_back(tpv);
					while (nums[st] == v1 && st < end)
					{
						st ++;
					}
					
					while (nums[end] == v2 && st < end)
					{
						end --;
					}
				}
				else if (nums[st] + nums[end] > anum)
				{
					end --;
				}
				else
				{
					st ++;
				}
			}
			
			while(nums[i] == num && i<len)
			{
				i ++;
			}
		}
		
		return result;
    }
};


/*
test cases:
invalid input: [] [1] [1 2]
solution: [-1 -2 -3 0 0 0 1 2 3], [-9 -9 3 6]
no solution: [1 2 3]
*/

File: array/15_3_sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums = sorted(nums)
        len_nums = len(nums)
        answer = []
        i = 0
        while i < len_nums - 2:
            if i > 0 and nums[i] == nums[i - 1]:
                i = i + 1
                continue
            j = i + 1
            k = len_nums - 1
            target = 0 - nums[i]
            while j < k:
                if nums[j] + nums[k] == target:
                    answer.append([nums[i], nums[j], nums[k]])
                    j = j + 1
                    k = k - 1
                    while j < k and nums[j] == nums[j - 1]:
                        j = j + 1
                    while j < k and nums[k] == nums[k + 1]:
                        k = k - 1
                elif nums[j] + nums[k] < target:
                    j = j + 1
                else:
                    k = k - 1
            i = i + 1
        return answer

s = Solution()
print s.threeSum([])
print s.threeSum([1,1,1])
print s.threeSum([0,0,0,0,0,0,0])
print s.threeSum([2,-1,-1])
print s.threeSum([2,-1,-1, -1, -1, -1])